### PR TITLE
copr: add ISO 8601 support for time parsing

### PR DESCRIPTION
**Pull request by ichn-hu**
### What problem does this PR solve?

Problem Summary: previously in TiDB https://github.com/pingcap/tidb/pull/20534 , we supported parsing time string in ISO 8601 format, but the change has not yet ported to TiKV, and this PR gets the job done.

### What is changed and how it works?

What's Changed:

### Related changes

- Need to cherry-pick to the release branch 4.0.

### Check List 

Tests 

- Unit test
- Integration test, see https://github.com/pingcap/tidb-test/pull/1149

Side effects

- Performance regression
    - Consumes more CPU

### Release note 

- port ISO8601 parsing from TiDB to TiKV's coprocessor.

### CODE REVIEW

File: components/tidb_query_datatype/src/codec/mysql/time/mod.rs [hunk 1]
```diff
@@ -369,11 +404,100 @@ mod parser {
             buffer = rest;
         }
 
+        let mut tz_offset = 0i32;
+        let mut tz_sign: &[u8] = b"".as_ref();
```
Comment (skyzh) [inline components/tidb_query_datatype/src/codec/mysql/time/mod.rs:408]: You can just use `let tz_sign: &[u8];` without giving an initial value. Rust compiler will check if it is assigned before using.

File: components/tidb_query_datatype/src/codec/mysql/time/mod.rs [hunk 2]
```diff
@@ -369,11 +404,100 @@ mod parser {
...
+        let mut tz_hour: &[u8] = b"".as_ref();
+        let mut tz_minute: &[u8] = b"".as_ref();
+        let mut has_tz = false;
+        // the following statement handles timezone
+        match components.len() {
+            9 => {
+                // 2020-12-23 15:59:23.233333+08:00
+                (separators.len() == 8).as_option()?;
+                match separators[6..] {
+                    [b"+", b":"] | [b"-", b":"] => {
```
Comment (skyzh) [inline components/tidb_query_datatype/src/codec/mysql/time/mod.rs:418]: Can we use `b"+:"` here?

**Review by skyzh — COMMENTED** [review]
Rest LGTM

**Comment by ichn-hu** [comment]
@skyzh PTALA

**Review by skyzh — APPROVED** [review]

**Comment by ichn-hu** [comment]
???

**Comment by ichn-hu** [comment]
WTF???

**Comment by LittleFall** [comment]
LMAO

**Pull request merged**